crates/ty_python_semantic/src/types/typed_dict.rs

```diff
@@ -76,6 +80,243 @@ impl<'db> TypedDictType<'db> {
+                                    )
+                                    .intersect(
+                                        db,
+                                        target_item_field.declared_ty.has_relation_to_impl(
+                                            db,
+                                            self_item_field.declared_ty,
+                                            inferable,
+                                            relation,
+                                            relation_visitor,
+                                            disjointness_visitor,
+                                        ),
+                                    )
+                            },
+                        )
+                    } else {
+                        // Self is missing this not-required, mutable field. This isn't ok if self
+                        // has read-only extra items, which all `TypedDict`s effectively do until
+                        // we support `closed` and explicit `extra_items`. See "A subtle
+                        // interaction between two structural assignability rules prevents
+                        // unsoundness" in `typed_dict.md`.
+                        // TODO: `closed` and `extra_items` support will go here.
+                        ConstraintSet::from(false)
+                    }
+                }
+            };
+            constraints.intersect(db, field_constraints);
+            if constraints.is_never_satisfied(db) {
+                return constraints;
+            }
+        }
+        constraints
+    }
+
+    // The only non-`TypedDict` that a `TypedDict` is assignable to is `Mapping[str, object]`.
+    // Although every instance of `TypedDict` is also an instance of `dict[str, object]`, a
+    // `TypedDict` still isn't assignable to `dict`, because mutating the `dict` could break
+    // `TypeDict`'s "structural" invariants.
+    // TODO: When we support `closed` and/or `extra_items`, we could allow assignments to other
+    // compatible `Mapping`s. `extra_items` could also allow for some assignments to `dict`, as
+    // long as `total=False`. (But then again, does anyone want a non-total `TypedDict` where all
+    // key types are a supertype of the extra items type?)
+    fn has_relation_to_non_typeddict_impl(
+        db: &'db dyn Db,
+        target: Type<'db>,
+    ) -> ConstraintSet<'db> {
+        debug_assert!(!matches!(target, Type::TypedDict(_)));
+        if let Some(specialization) = target.known_specialization(db, KnownClass::Mapping)
+            && let &[key_ty, value_ty] = specialization.types(db)
+            // The key type must be exactly `str`.
+            && key_ty == KnownClass::Str.to_instance(db)
+            // The value type must be `object` (or a gradual type that's consistent with `object`).
+            && KnownClass::Object
+                .to_instance(db)
+                .is_assignable_to(db, value_ty)
```

Comment:
Is this the right way to do these tests? Do I need to `.normalize()` the `str` class?